src/mpi/request/test.c

```diff
@@ -64,23 +52,9 @@ int MPIR_Test_impl(MPI_Request * request, int *flag, MPI_Status * status)
     }
 
     if (MPIR_Request_is_complete(request_ptr)) {
-        mpi_errno = MPIR_Request_completion_processing(request_ptr, status, &active_flag);
-        if (!MPIR_Request_is_persistent(request_ptr)) {
-            MPIR_Request_free(request_ptr);
-            *request = MPI_REQUEST_NULL;
-        }
-        if (mpi_errno)
-            MPIR_ERR_POP(mpi_errno);
         *flag = TRUE;
```

Comment:
What's the point of this flag?  All it's doing is checking if `MPIR_Request_is_complete` is true.  Whoever is calling this function can just check that instead of checking the flag.

Reply:
The flag is the user's provided flag passed from upper layer, either this will set this, or the upper layer has to set it. I guess it boils down to what we want the "device" layer to do.

Reply:
It makes sense for the MPI interface, since the user has no other way of checking for completion.  But it doesn't make sense for the MPID interface (or in this case, internal to the MPIR layer), since the `MPIR_Request_is_complete` function exists to do that.

Reply:
What do you suggest for the device layer then ?

Reply:
You refer to the "device layer", which is a bit confusing.  In the patch I'm talking about, both functions are in the MPI layer.  And one function is announcing the completion of the request in two ways: through the request itself and through the flag.  I'm just saying that it's redundant.  One of them is sufficient.